sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,62 +32,109 @@
 
 import org.joda.time.Instant;
 
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+
 /**
  * Determine the timing and other properties of a new pane for a given computation, key and window.
  * Incorporates any previous pane, whether the pane has been produced because an
  * on-time {@link AfterWatermark} trigger firing, and the relation between the element's timestamp
  * and the current output watermark.
  */
 public class PaneInfoTracker {
+  @VisibleForTesting
+  static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
+      StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
+
   private TimerInternals timerInternals;
 
   public PaneInfoTracker(TimerInternals timerInternals) {
     this.timerInternals = timerInternals;
   }
 
-  @VisibleForTesting
-  static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
-      StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
+  /**
+   * Could we have any previous pane info stored?
+   */
+  public boolean hasState(ReduceFn<?, ?, ?, ?>.Context context) {
+    return context.window().maxTimestamp().isBefore(GlobalWindow.INSTANCE.maxTimestamp());
+  }
 
-  public void clear(StateAccessor<?> state) {
-    state.access(PANE_INFO_TAG).clear();
+  /**
+   * Clear any pane info state.
+   */
+  public void clear(ReduceFn<?, ?, ?, ?>.Context context) {
+    if (hasState(context)) {
+      context.state().access(PANE_INFO_TAG).clear();
+    }
+    // else: nothing to clear if for GlobalWindow.
   }
 
   /**
    * Return a ({@link ReadableState} for) the pane info appropriate for {@code context}. The pane
    * info includes the timing for the pane, who's calculation is quite subtle.
    *
    * @param isFinal should be {@code true} only if the triggering machinery can guarantee
-   * no further firings for the
+   *                no further firings for the trigger.
    */
   public ReadableState<PaneInfo> getNextPaneInfo(
       ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal) {
-    final Object key = context.key();
-    final ReadableState<PaneInfo> previousPaneFuture =
-        context.state().access(PaneInfoTracker.PANE_INFO_TAG);
-    final Instant windowMaxTimestamp = context.window().maxTimestamp();
-
-    return new ReadableState<PaneInfo>() {
-      @Override
-      public ReadableState<PaneInfo> readLater() {
-        previousPaneFuture.readLater();
-        return this;
-      }
-
-      @Override
-      public PaneInfo read() {
-        PaneInfo previousPane = previousPaneFuture.read();
-        return describePane(key, windowMaxTimestamp, previousPane, isFinal);
-      }
-    };
+    if (hasState(context)) {
```

Comment:
Since you are re-writing this anyhow, can you make it explicitly parameterized by the pieces of the context it really needs?